File: 백준/Silver/11047. 동전 0/동전 0.py
# 동전의 개수, 총합을 입력 받음.
# 동전을 배열에 넣은 뒤 가장 큰 동전부터 차례대로 계산하면 최적의 값을 찾을 수 있음.
import sys

N, K = map(int, sys.stdin.readline().split())
coin = []
for _ in range(N):
  coin.append(int(sys.stdin.readline()))

# 가장 큰 값 부터 뽑아내기 위해서  
coin.sort(reverse=True)

# 총 동전의 개수
count = 0

for i in range(N):
  if K >= coin[i]:
    temp = K // coin[i]
    count += temp
    K %= coin[i]
    
print(count)

File: 백준/Silver/11047. 동전 0/동전 0.java

import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

		String[] sa = bf.readLine().split(" ");
		int N = Integer.parseInt(sa[0]);
		int K = Integer.parseInt(sa[1]);

		int[] coins = new int[N];
		for (int i = 0; i < N; i++) {
			coins[i] = Integer.parseInt(bf.readLine());
		}

		int coinCnt = 0;
		int curCoin = 0;
		for (int i = N - 1; i >= 0; i--) {
			int coin = coins[i];

			if(coin > K - curCoin) continue;

			int a = (K - curCoin) / coin;
			coinCnt += a;
			curCoin = curCoin + (coin * a);
		}

		System.out.println(coinCnt);

	}
}
